src/main/java/com/eerussianguy/blazemap/feature/maps/AtlasExporter.java

```diff
@@ -0,0 +1,144 @@
+                    for(int regionZ = atlasInfo.atlasStartZ; regionZ <= atlasInfo.atlasEndZ; regionZ++) {
+                        File file = new File(folder, LayerRegionTile.getImageName(new RegionPos(regionX, regionZ)));
+                        if(!file.exists()) continue;
+                        NativeImage tile = NativeImage.read(Files.newInputStream(file.toPath()));
+
+                        int regionOffsetX = (regionX - atlasInfo.atlasStartX) * resolution.regionWidth;
+                        int regionOffsetZ = (regionZ - atlasInfo.atlasStartZ) * resolution.regionWidth;
+                        for(int x = 0; x < resolution.regionWidth; x++) {
+                            for(int z = 0; z < resolution.regionWidth; z++) {
+                                int atlasPixelX = regionOffsetX + x;
+                                int atlasPixelZ = regionOffsetZ + z;
+                                int atlasPixel = atlas.getPixelRGBA(atlasPixelX, atlasPixelZ);
+                                int tilePixel = tile.getPixelRGBA(x, z);
+                                atlas.setPixelRGBA(atlasPixelX, atlasPixelZ, Colors.layerBlend(atlasPixel, tilePixel));
+                            }
+                        }
+
+                        // non-atomic op on volatile int is ok because only 1 thread writes to variable
+                        // Java guarantees r/w access to 32-bit variables is atomic, so there are no mid-write reads
+                        task.tilesCurrent++;
+                    }
+                }
+            }
+
+            File file = new File(Minecraft.getInstance().gameDirectory, "screenshots/blazemap-export-"+ (System.currentTimeMillis() / 1000L) +".png");
+            file.getParentFile().mkdirs();
+            atlas.writeToFile(file);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        task = null;
+    }
+
+    private static NativeImage constructAtlas(StorageAccess.Internal storage, TileResolution resolution, AtlasInfo atlasInfo) {
+        int minX = Integer.MAX_VALUE, maxX = Integer.MIN_VALUE, minZ = Integer.MAX_VALUE, maxZ = Integer.MIN_VALUE;
+        int regionsX, regionsZ;
+
+        var layers = task.map.value().getLayers();
+        for(var layer : layers) {
+            if(!task.layers.contains(layer)) continue;
+            File folder = storage.getMipmap(layer.location, ".", resolution);
+            File[] images = folder.listFiles();
+            if(images == null) continue;
+
+            for(var image : images) {
+                String[] coords = image.getName().replaceAll("(^\\[)|(]\\.png$)", "").split(",");
+                if(coords.length != 2) continue;
+                // non-atomic op on volatile int is ok because only 1 thread writes to variable
+                // Java guarantees r/w access to 32-bit variables is atomic, so there are no mid-write reads
+                task.tilesTotal++;
```

Comment:
Generally, the assumption stated above shouldn't be taken for a `var++` action, as in byte code that's actually three operations:
- read var
- add 1 to var
- write new value to var

However, since there should only ever be a single thread accessing each task, that's fine here!

Not essential for right now, but this might be a good reason to have an object per export and perform these functions non-statically on that object. That way, you can use scoping of variables (`private`, `protected`) to prevent anybody else from changing them! And as long as the object doing the export only ever exists on that one thread, and is cleaned up once its job is done, you can completely ignore practically all thread-safety concerns as there's no way for any other thread to possibly mess up your data 😊

Reply:
Ok so bottom comment is wrong in this context.
Will clean that up.

Reply:
Actually, no, I will clarify the comment.

We are only writing from one thread. And reading from the work thread (that also writes) and the render thread.
The point of this comment is that, because the int is written in one shot, the render thread will always either read the old or new value, and never a value corrupted by tearing. This means the worst thing the lack of synchronization does here is let our old value lag for one frame (which with a race condition, we could still have)

Reply:
Ah gotcha. That wasn't super clear from the comment. But no worries, all good then!

Reply:
Fixed in working branch